gtk/src/toga_gtk/window.py

```diff
@@ -143,11 +169,89 @@ def hide(self):
     # Window state
     ######################################################################
 
-    def set_full_screen(self, is_full_screen):
-        if is_full_screen:
+    def get_window_state(self):
+        window_state_flags = self._window_state_flags
+        if window_state_flags & Gdk.WindowState.MAXIMIZED:
+            return WindowState.MAXIMIZED
+        elif window_state_flags & Gdk.WindowState.ICONIFIED:
+            return WindowState.MINIMIZED  # pragma: no-cover-if-linux-wayland
+        elif window_state_flags & Gdk.WindowState.FULLSCREEN:
+            if self._in_presentation:
+                return WindowState.PRESENTATION
+            elif self._in_fullscreen:
+                return WindowState.FULLSCREEN
+            else:
+                return WindowState.NORMAL
+        else:
+            return WindowState.NORMAL
+
+    def set_window_state(self, state):
+        if ("WAYLAND_DISPLAY" in os.environ) and (
```

Comment:
pardon the interjection...but I was curious if there has been discussions around this methodology to detect which protocol the display is using?

While the environment variable `WAYLAND_DISPLAY` will probably be a reliable indicator that wayland is in use, this seems inherently fragile; I mean, it's trivial to set environment variables.

Furthermore, though, I can force an app to use X11 via Xwayland:
```
toga/examples/tutorial1 on  main [!] via 🐍 v3.12.3 (briefcase-3.12) took 38s 
❯ GDK_BACKEND=x11 python -m tutorial
Gdk.Display.get_default()=<GdkX11.X11Display object at 0x7f29e6da6500 (GdkX11Display at 0x35d826c0)>
^C

toga/examples/tutorial1 on  main [!] via 🐍 v3.12.3 (briefcase-3.12) took 4s 
❯ GDK_BACKEND=wayland python -m tutorial
Gdk.Display.get_default()=<__gi__.GdkWaylandDisplay object at 0x7f2fe8cfee40 (GdkWaylandDisplay at 0xde391c0)>
```

I've spent a little time seeing if PyGObject is exposing any of the useful C macros that can evaluate the nature of the display....not to much avail, though. But that's mostly why I'm back to my original question in case we've already evaluated this :)

Reply:
Thanks, I appreciate it and also wanted to discuss about it. While implementing the screens API, I had found 3 main methods to detect Wayland:

1. Check for the presence of Wayland sockets in `/run/user/$UID/wayland-*`
2. Check if xrandr works i.e., `xrandr --listmonitors`
3. Check for environment variables like: `$WAYLAND_DISPLAY`, `$XDG_SESSION_TYPE` - Both of them are equally reliable, but  `$WAYLAND_DISPLAY` is simpler to check.

The 2nd one seemed way more fragile & unreliable.  

The first one seemed a bit more compositor-implementation specific thing, as I am not sure if the presence of a Wayland socket at `/run/user/$UID/` is defined in the Wayland protocol specification. 

Hence, I had opted to use the third option. I know that it seems a bit fragile but I have manually tested the behaviour on 3 different DEs - KDE Plasma, Gnome, Cinnamon and on 5 different distros. The behaviour of `$WAYLAND_DISPLAY` was as expected and gave the same results. So, I went ahead with it. 

Let me know, what do you think about it :)

Reply:
I dug in to this a bit more today. Ideally, I think, Toga would be abstracted away from the details of X11 and Wayland altogether; however, obviously, there are high-level consequences when we aren't using X11 and we need a way to detect this situation.

I don't think Toga should be trying to make this determination itself; instead, we should defer to the GNOME stack we're building on since it's intimately aware of these details and already managing all of this. Since it was not at all apparent how to extract this information with PyGObject, I eventually [solicited](https://discourse.gnome.org/t/detecting-wayland-verses-x11-with-pygobject/22858) the GNOME team for some insight.

While we remain on Gtk3, we can do something like this:
```python
import gi
gi.require_versions({'Gdk': '3.0', 'Gtk': '3.0', 'GdkX11': '3.0'})
from gi.repository import Gdk, Gtk, GdkX11

IS_WAYLAND = not isinstance(Gdk.Display.get_default(), GdkX11.X11Display)
```

Once we're able to transition to Gtk4, we can make this a little more straightforward but it wouldn't be required:
```python
import gi
gi.require_versions({'Gdk': '4.0', 'Gtk': '4.0', 'GdkWayland': '4.0'})
from gi.repository import Gdk, Gtk, GdkWayland

IS_WAYLAND = isinstance(Gdk.Display.get_default(), GdkWayland.WaylandDisplay)
```
On Gtk3, `GtkWayland` is not exposed through the introspection....but in Gtk4, it is.

We should be able to tuck this away somewhere and replace the `os.environ` refs with a `IS_WAYLAND` global.

Reply:
One last note: this code may not be generally portable. However, Toga only uses Gtk in *NIX environments where a `Display` must use X11 or Wayland; so, it's portable insofar as Toga is concerned.

Reply:
I agree, this seems like a good option. I think we can put this in `core/src/platform.py` or any other place you deem appropriate. But, I think this PR is already huge and I will leave it to you or someone else to implement it :)

As a side note, when we transition to Gtk4, I don't think this window state API, or even the screens API would work properly, since gtk4 has made much of the current API non-functional or has downright removed them. Wayland's security features are making this even more difficult. Gtk's docs mention one thing and their APIs do another thing.

Even while implementing this PR, I have found wayland doesn't allow for proper minimization-restoration, full screening, etc. and even the maximization-restoration gtk APIs fails sometimes. I am thinking of disabling window state setting altogether on wayland, as the behavior seems highly fragile and is breaking in new ways with each new release of gnome.

Hence, I think I'll just avoid implementing for Wayland and raise a `NotImplementedError`. Let me know what you think.